docs/manual/scala/guide/cluster/DomainModelling.md

```diff
@@ -1,19 +1,202 @@
+```
+
+### Defining Commands Handlers
+
+Once you define your protocol in terms of Commands, Replies, Events and State, you need to specify the business rules of your model. The command handlers define how to handle each incoming command, which validations must be applied, and finally, which events will be persisted if any.
+
+You can encoding it in different ways. The [recommended style](https://doc.akka.io/docs/akka/2.6/typed/persistence-style.html#command-handlers-in-the-state) is to add the command handlers in your state classes. Since `ShoppingCart` have two state classes extensions, it makes sense to add the repective business rules validation on each state class. Each possible state will define how each command should be handled.
+
+```scala
+sealed trait ShoppingCart  {
+  def applyCommand(cmd: ShoppingCartCommnad): ReplyEffect[ShoppingCartEvent, ShoppingCartState]
+}
+
+case class OpenShoppingCart(items: Map[String, Int]) extends ShoppingCart {
+  def applyCommand(cmd: ShoppingCartCommnad) =
+    cmd match {
+      case UpdateItem(_, qty, replyTo) if qty < 0 =>
+        Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
+
+      // an item is delete by setting it's quantity to 00
+      case UpdateItem(productId, 0, replyTo) if !items.contains(productId) =>
+        Effect.reply(replyTo)(Rejected("Cannot delete item that is not already in cart"))
+
+      case UpdateItem(productId, quantity, replyTo) =>
+        Effect
+          .persist(ItemUpdated(productId, quantity))
+          .thenReply(replyTo) { updatedCart => // updated cart is state updated after applying ItemUpdated
+            Accepted
+          }
+
+      // check it out
+      case Checkout(replyTo) =>
+        Effect
+          .persist(CheckedOut)
+          .thenReply(replyTo){ updatedCart => // updated cart is state updated after applying CheckedOut
+            Accepted
+          }
+
+      case Get(replyTo) =>
+        Effect.reply(replyTo)(CartState(items, status = "open"))
+  }
+}
+
+case class CheckedOutShoppingCart(items: Map[String, Int]) extends ShoppingCart {
+  def applyCommand(cmd: ShoppingCartCommnad) =
+    cmd match {
+      // CheckedOut is a final state, not mutations allowed
+      case UpdateItem(_, _, replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot update a checked-out cart"))
+      // CheckedOut is a final state, not mutations allowed
+      case Checkout(replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+
+      // it is allowed to read it's state though
+      case Get(replyTo) =>
+        Effect
+        .reply(replyTo)(CartState(items, status = "checked-out"))
+    }
+
+}
+```
+
+Command handlers are the meat of the model. They encode the business rules of your model and act as a guardian of the model consistency. Any mutation must be validated by it. However, they don't apply the mutations. Instead, they express the mutations in the form of persisted events.
```

Comment:
"Any mutation must be validated by it. However, they don't apply the mutations. Instead, they express the mutations in the form of persisted events."

This was a bit confusing. How about:

```suggestion
Command handlers are the meat of the model. They encode the business rules of your model and act as a guardian of the model consistency. Command handler must first validate the incoming command can mutate the current state into a valid new state. However, the command handler will not produce the mutation and, instead, the command handler will express the mutations in the form of persisted events.
```